src/java.base/share/classes/sun/security/rsa/RSAKeyPairGenerator.java

```diff
@@ -123,70 +150,121 @@ public void initialize(AlgorithmParameterSpec params, SecureRandom random)
         }
 
         this.keySize = tmpKeySize;
-        this.publicExponent = tmpPublicExponent;
-        this.random = random;
+        this.publicExponent = tmpPubExp;
+        this.random = (random == null? JCAUtil.getSecureRandom() : random);
+        this.useNew = useNew;
     }
 
-    // generate the keypair. See JCA doc
+    // FIPS 186-4 B.3.3 / FIPS 186-5 A.1.3
+    // Generation of Random Primes that are Probably Prime
     public KeyPair generateKeyPair() {
-        // accommodate odd key sizes in case anybody wants to use them
-        int lp = (keySize + 1) >> 1;
+        BigInteger e = publicExponent;
+        BigInteger minValue = (useNew? getSqrt(keySize) : ZERO);
+        int lp = (keySize + 1) >> 1;;
         int lq = keySize - lp;
-        if (random == null) {
-            random = JCAUtil.getSecureRandom();
+        int pqDiffSize = lp - 100;
+
+        BigInteger p = null;
+        int i = 0;
+        while (i++ < 10*lp) {
+            BigInteger tmpP = BigInteger.probablePrime(lp, random);
+            if ((!useNew || tmpP.compareTo(minValue) == 1) &&
+                    isRelativePrime(e, tmpP.subtract(ONE))) {
+                p = tmpP;
+                break;
+            }
         }
-        BigInteger e = publicExponent;
-        while (true) {
-            // generate two random primes of size lp/lq
-            BigInteger p = BigInteger.probablePrime(lp, random);
-            BigInteger q, n;
-            do {
-                q = BigInteger.probablePrime(lq, random);
-                // convention is for p > q
-                if (p.compareTo(q) < 0) {
-                    BigInteger tmp = p;
-                    p = q;
-                    q = tmp;
+        if (p == null) {
+            throw new ProviderException("Cannot find prime P");
+        }
+
+        BigInteger q = null;
+        i = 0;
+
+        while (i++ < 20*lq) {
+            BigInteger tmpQ = BigInteger.probablePrime(lq, random);
+
+            if ((!useNew || tmpQ.compareTo(minValue) == 1) &&
+                    (p.subtract(tmpQ).abs().compareTo
+                            (TWO.pow(pqDiffSize)) == 1) &&
+                    isRelativePrime(e, tmpQ.subtract(ONE))) {
+                q = tmpQ;
+                BigInteger n = p.multiply(q);
+                if (!useNew && n.bitLength() != keySize) {
+                    // regenerate Q if n is not the right length
+                    continue;
+                }
+                KeyPair kp = createKeyPair(type, keyParams, n, e, p, q);
+                if (kp != null) {
+                    return kp;
                 }
```

Comment:
The  logic may be more clear if moving the checking of n and key generation out of the loop for q, by regenerate both p and q if needed.

Reply:
Ok, will add an outer while-loop as in the existing code.